src/stackmap_checker/call_stack_state.c

```diff
@@ -4,125 +4,142 @@
+        frames[depth - 1].opt_ret_addr =
+            *(uint64_t *)(frames[depth - 1].registers[UNW_X86_64_RBP] + 8);
+        // Store the current BP.
+        frames[depth - 1].bp = frames[depth - 1].real_bp =
+            frames[depth - 1].registers[UNW_X86_64_RBP];
+        frames[depth - 1].inlined = 0;
+        // Stop when main is reached.
         char fun_name[MAX_BUF_SIZE];
         unw_get_proc_name(&cursor, fun_name, sizeof(fun_name), &off);
-        // Stop when main is reached.
         if (!strcmp(fun_name, "main")) {
+            state->main_regs = get_registers(cursor);
+            state->main_bp = state->main_regs[UNW_X86_64_RBP];
+            state->main_ret_addr = (uint64_t)(state->main_bp + 8);
             break;
         }
-        if (!registers[frame]) {
-            // 16 registers are saved for each frame
-            registers[frame] = calloc(16, sizeof(unw_word_t));
-        }
-        unw_get_reg(&cursor, UNW_X86_64_RAX, &registers[frame][0]);
-        unw_get_reg(&cursor, UNW_X86_64_RDX, &registers[frame][1]);
-        unw_get_reg(&cursor, UNW_X86_64_RCX, &registers[frame][2]);
-        unw_get_reg(&cursor, UNW_X86_64_RBX, &registers[frame][3]);
-        unw_get_reg(&cursor, UNW_X86_64_RSI, &registers[frame][4]);
-        unw_get_reg(&cursor, UNW_X86_64_RDI, &registers[frame][5]);
-        unw_get_reg(&cursor, UNW_X86_64_RBP, &registers[frame][6]);
-        unw_get_reg(&cursor, UNW_X86_64_RSP, &registers[frame][7]);
-        unw_get_reg(&cursor, UNW_X86_64_R8,  &registers[frame][8]);
-        unw_get_reg(&cursor, UNW_X86_64_R9,  &registers[frame][9]);
-        unw_get_reg(&cursor, UNW_X86_64_R10, &registers[frame][10]);
-        unw_get_reg(&cursor, UNW_X86_64_R11, &registers[frame][11]);
-        unw_get_reg(&cursor, UNW_X86_64_R12, &registers[frame][12]);
-        unw_get_reg(&cursor, UNW_X86_64_R13, &registers[frame][13]);
-        unw_get_reg(&cursor, UNW_X86_64_R14, &registers[frame][14]);
-        unw_get_reg(&cursor, UNW_X86_64_R15, &registers[frame][15]);
-        // Store the address of the return address.
-        ret_addrs[frame] =
-            (uint64_t)(registers[frame][UNW_X86_64_RBP] + 8);
-        // Store the current BP.
-        *(bps + frame) = registers[frame][UNW_X86_64_RBP];
-        frame++;
     }
-    call_stack_state_t *state = malloc(sizeof(call_stack_state_t));
-    state->ret_addrs = ret_addrs;
-    state->bps       = bps;
-    state->registers = registers;
-    state->depth     = frame;
-    state->records   = NULL;
+    state->frames = frames;
+    state->depth  = depth;
     return state;
 }
 
 void collect_map_records(call_stack_state_t *state, stack_map_t *sm)
 {
-    state->records = calloc(state->depth - 1, sizeof(stack_map_record_t));
-    // Loop `depth - 1` times, because the return address of 'trace' should not
-    // be overwritten (it should still return in main)
-    for (int i = 0; i < state->depth - 1; ++i) {
+    for (size_t i = 0; i + 1 < state->depth; ++i) {
```

Comment:
Is `i + 1 < state->depth` better as `i <= state->depth`?

Reply:
No, it should actually be `i < state->depth - 1`, or `i <= state->depth - 2`, but `i` is a `size_t`.